### DISCUSSION TITLE
Sort out `bitcoincore-rpc`

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2777] [General]
It has recently come to light that `rust-bitcoincore-rpc` is being used in the wild because `electrs` depends on it. Until now it was my assumption that it was only used in integration tests. 

## My claims

- The `rust-bitcoincore-rpc` crate is not production ready (justified by the fact that the latest version tested against is Bitcoin Core `v0.21.0`.
- There is no-one in the `rust-bitcoin` dev community that wants to seriously maintain a JSON-RPC client. 

## Client conundrum

- It seems reasonable that `rust-bitcoin` should provide the JSON-RPC types for Bitcoin Core.
- It seems unreasonable that `rust-bitcoin` should provide a battle tested, secure, production ready, JSON-RPC client.

However in order to test the types we have to implement a client.

## Possible solution

I have made an initial attempt at a solution: ~https://github.com/tcharding/rust-bitcoin-core-json-rpc~ https://github.com/tcharding/rust-bitcoind-json-rpc I'm unsure where the code in this solution should live, I think at a minimum the JSON types should be re-exported from the `rust-bitcoin` crate.

To do the rest of the types as I've done is a non-trivial amount of work so I'm chasing feedback please on whether this is valuable and if its a good approach. (Even though we can copy a lot of it from the `rust-bitcoincore-rpc-json`.)

As I've said before, I am not a web hacker and I know basically nothing about JSON-RPC. Any and all suggestions most appreciated.

**Comment by tcharding** [comment]
cc @thomaseizinger because he is the guy I know who knows most about web stuff (and has an interest in it) and is a Rust hacker.

**Reply by thomaseizinger** [reply]
I am not invested in into how `rust-bitcoin` solves their need for a JSON-RPC client but in general, I think it would be best to define the API once and leave it up to the end-user, which HTTP client they end up using to actually send their requests. https://github.com/thomaseizinger/rust-jsonrpc-client works this way but there is no need to take that dependency, what it generates can also easily be vendored.

Only concerning yourself with the API definition in `rust-bitcoin` (speaking about the org here) has several advantages:

- It takes away all of the annoying boilerplate of defining the JSON-RPC interface to Bitcoin Core.
- You avoid writing an HTTP + JSON-RPC client.

IMO the two things above form a sweet-spot in terms of maintenance effort vs utility for end users.

Defining all the payload types is what actually takes time and is the bit that you don't want to duplicate across projects so it makes sense that it is defined in a single place. Writing an HTTP client is what you _don't_ want to do because it just attracts more work as people come and ask for features ("I want it to work on async-std", "I need TLS", ...). It is also completely irrelevant for JSON-RPC itself so any client will do. I'd assume that this isn't exactly the stuff that the maintainers of `rust-bitcoin` want to spend their time on.

So yeah, I'd disagree with your task list above. I think you don't need to add async support to any client. Once you define a trait for sending an HTTP POST request, you can build out the entire JSON-RPC API of Bitcoin Core without taking any dependency on any HTTP infrastructure. That will make the resulting crate super lightweight and users can plug in whatever HTTP client they see fit.

**Reply by tcharding** [reply]
Legendary, thanks @thomaseizinger, glad I mentioned you.

**Reply by apoelstra** [reply]
I like this approach. It is something of an inversion of our current approach, where we have rust-jsonrpc doing a bunch of HTTP stuff around a generic transport, and then bitcoin-core-rpc calling this non-generically.

Instead, bitcoin-core-rpc would provide a trait defining an arbitrary RPC endpoint (not necessarily HTTP or anything, not caring about authentication, not caring about other headers or encryption or blah blah blah), and then using that trait to implement a well-defined "request this, get that" API. Maybe, this trait has a method to define the version of the RPC interface or something.

The `bitcoind` crate could then be an implementor of this trait; it stands up a specific version of bitcoind, finds a spare port to put the RPC server on, and then uses `minreq` to talk to the server. Or something.

**Reply by dpc** [reply]
> Defining all the payload types is what actually takes time and is the bit that you don't want to duplicate across projects so it makes sense that it is defined in a single place.

The payloads are already collected in their own crate for this very reason
https://github.com/rust-bitcoin/rust-bitcoincore-rpc/tree/master/json

**Reply by tcharding** [reply]
I'm not exactly sure I understand everything above but I made start on testing the types in `json`. If we can at least do that and test each type against various versions of Core I think that is at least progress. Holla at me if I've got things twisted please.

https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/358

**Reply by yancyribbens** [reply]
I think the discussion is about using an api schema that is independent from the mechanism used to serve the content.  I'm not super familiar with this project though, just my interpretation of the comments above.

**Reply by romanz** [reply]
Indeed, https://github.com/romanz/electrs depends on https://github.com/rust-bitcoin/rust-bitcoincore-rpc - many thanks to all of its maintainers!

> * It seems reasonable that `rust-bitcoin` should provide the JSON-RPC types for Bitcoin Core.
> * It seems unreasonable that `rust-bitcoin` should provide a battle tested, secure, production ready, JSON-RPC client.

> Instead, bitcoin-core-rpc would provide a trait defining an arbitrary RPC endpoint (not necessarily HTTP or anything, not caring about authentication, not caring about other headers or encryption or blah blah blah), and then using that trait to implement a well-defined "request this, get that" API. Maybe, this trait has a method to define the version of the RPC interface or something.

I definitely agree - and it would be great to continue using the JSON-RPC types.
Since electrs is not using async Rust, it would be fine for me to use [minreq](https://github.com/neonmoe/minreq)-based client if needed.

**Reply by Kixunil** [reply]
@apoelstra while I'd love to see central definitions of Core methods with their inputs and outputs, I strongly suspect it'd require async-in-traits and we'd need both sync and async versions. It'd be nice if we could workaround it.

**Comment by yancyribbens** [comment]
I haven't looked too close, although curiously the blockstream fork of electrs https://github.com/Blockstream/electrs does not use list `bitcoincore-rpc` as a dep in Cargo.toml.

**Reply by tcharding** [reply]
Interestingly they have it as a transient `dev-dependency` (via `bitcoind`) but not in production.

**Reply by shesek** [reply]
Correct, `bitcoincore-rpc` is dependent on via `bitcoind` and also used for [the tests](https://github.com/Blockstream/electrs/blob/new-index/tests/common.rs), but not for the main `Daemon` implementation.

But we were just recently discussing moving away from the custom RPC client implementation (originally inherited from `romanz/electrs` before they switched) to `bitcoincore-rpc`. Was planning to look into this shortly, but now I wonder if I should wait for the "do it right version"? fwiw `async` isn't really necessary for this.

**Reply by tcharding** [reply]
Cheers, I should have something for you soon to at least evaluate.

**Reply by tcharding** [reply]
Hey @shesek, if you have time can you please check out the updated initial post above. Also, as a potential consumer, could you please answer. something for me? 

If we provide the JSON-RPC types (tested against for example Core v22 through to 27) and also a bunch of clients that are documented to be only provided for integration testing would you write your own RPC client or would you guys likely (ab)use one of the integration-testing clients in production?

I'm not trying to throw you under the bus, its a real question.

**Reply by shesek** [reply]
I can't answer for Blockstream, but as far as I'm concerned I wouldn't use a library in a way that's explicitly against its maintainers' wishes.

I would, however, probably keep using the original `rust-bitcoincore-rpc`, whether its under the `rust-bitcoin` umbrella or maintained elsewhere. I find it valuable despite its shortcoming, and it will most likely not go away - `electrs` and [`bdk`](https://docs.rs/bdk_bitcoind_rpc) use it and will probably have interest in maintaining it as a general-purpose RPC client library.

I'm also already using in a couple projects of mine (Bitcoin Wallet Tracker and another unpublished one) and don't really have another alternative to switch to.

> There is no-one in the `rust-bitcoin` dev community that wants to seriously maintain a JSON-RPC client.

Have you considered checking with authors of projects using `rust-bitcoincore-rpc`? Perhaps @romanz or the @bitcoindevkit folks have interest in maintaining it if you offered them commit access?

**Reply by dpc** [reply]
Seconded. We're using in Fedimint, we don't mind no async support (we can wrap it with `block_in_place`/`spawn_blocking`) and is simple, trusted and does the job.

**Reply by tcharding** [reply]
Correct me if I'm wrong but what you guys want is a simple RPC client (no async) that does what you want and you don't want to have to change anything, things are fine just now. I think I can deliver whats in my head _and_ satisfy you (if I've understood correctly). Thanks for the responses, please do say if I've misunderstood.

**Reply by dpc** [reply]
I guess my point is that the current `rust-bitcoincore-rpc` is OK and works in a lot of applications, and lack of `async` is not a big deal. If there will be something "better", we might switch, but it never felt like a pressing matter.

**Reply by romanz** [reply]
> I guess my point is that the current rust-bitcoincore-rpc is OK and works in a lot of applications, and lack of async is not a big deal. If there will be something "better", we might switch, but it never felt like a pressing matter.

+1 for electrs :)

**Reply by shesek** [reply]
> Correct me if I'm wrong but what you guys want is a simple RPC client (no async) that does what you want and you don't want to have to change anything, things are fine just now.

Well, ideally it would also be well maintained and kept up with newer versions of Bitcoin Core. :) But basically yes. And I fully understand if maintaining a general-purpose RPC client is outside the rust-bitcoin scope.

**Comment by tcharding** [comment]
Instead of messing with `bitcoincore-rpc-json` I thought I might just hack up a new crate that does:

- Provides structs for _all_ Bitcoin Core JSON RPC calls.
- Tests against every version of Bitcoin Core starting at `v22.0`.

I started by just implementing `getblockchaininfo` and writing a test that uses `jsonrpc-client` against Bitcoin Core `v22.0`. Can I get some feed back please as to whether this is useful or if I'm wasting my time. If I get a concept ACK I'll start adding all the other RPC calls.

https://github.com/tcharding/rust-bitcoin-core-json-rpc-types

(Should move to this org if successful.)

Notes:

- I'm stealing the structs from `bitcoincore-rpc-json`
- I'm using a slightly different approach, providing a `json` module that provides types without using `rust-bitcoin` (ie using stdlib types, and some custom types) then providing a `concrete` module that provides types of the same name with `rust-bitcoin` types. This hopefully helps with:
   - finer control of the conversion
   - better error handling
   - easier rust-bicoin upgrate path
   - maybe debugging, when we start hitting differences in the different Core version RPC APIs.

Thanks in advance!

**Reply by tcharding** [reply]
I'm still not totally sure how the post trait discussed above will look but I'm thinking the new crate can define all the payloads POST and GET then any additional interface can come later.

**Reply by apoelstra** [reply]
This sounds like a good approach. Unsure how exactly you will handle `Amount`s in your "stdlib types only" implementation. I glanced at your code and it currently avoids this issue, but what you have implemented seems fine.

>I'm still not totally sure how the post trait discussed above will look but I'm thinking the new crate can define all the payloads POST and GET then any additional interface can come later.

This crate shouldn't know or care about the distinction between POST and GET.

**Reply by tcharding** [reply]
Cool, cheers. I'll flesh it out a bit and pick an API call with an `Amount` to see how it looks.

**Reply by romanz** [reply]
Concept ACK - many thanks @tcharding for the help!
Will review https://github.com/tcharding/rust-bitcoin-core-json-rpc/commit/9f7ad5a860bf26d24a3cbe36abfe665c0b8ac3cc soon.

**Reply by tcharding** [reply]
Now living at: https://github.com/tcharding/rust-bitcoind-json-rpc

**Comment by tcharding** [comment]
Thanks for everyones input so far, I've update the initial post. Call for Feeback please.

**Comment by tcharding** [comment]
Seems this problem has been around for a long time: https://github.com/rust-bitcoin/rust-bitcoin/issues/184

**Comment by tcharding** [comment]
I'd like to present what I've been hacking on for the last few days: https://github.com/tcharding/rust-bitcoind-json-rpc

It is an attempt to provide JSON-RPC types for multiple versions of Bitcoin Core and exhaustively test them implementing simple per-version RPC clients using `jsonrpc`.

It uses macros to implement and test every API method separately so one can see exactly what has been implemented and what has been tested per `bitcoind` version. Its a single commit although I iterated a bunch of times :)

I'd love some feedback, doing _all_ the rest of the RPC methods is a whole lot of work, would be mad to get some feedback before I plunge in.

In particular I've split the json types into two modules `json` and `model` - this was done on a gut feeling, RFC please. 

Thanks in advance.

cc @dpc and @thomaseizinger (for their general interest in interesting design problems and respect for their knowledge).

**Reply by apoelstra** [reply]
Looks good! I am a bit suspicious of parsing amounts as f64s but if the expected application is to convert them immediately to an `Amount` without attempting to do any arithmetic on them, I think it's fine. Whenever we explain the division between the raw `json` crate/module and the one that has the rust-bitcoin types, we should clarify that.

It used to be that we couldn't even parse amounts as `f64`s using serde_json. We had a long saga of weird BS using the strason crate, rust_decimal, the `arbitrary_precision` feature of serde_json, and so on. But I am fuzzing right now to find a f64 in the Bitcoin range that serde_json won't round-trip and coming up clean, which is great.

TBH I don't really understand the point of all these macros. It seems like they are wrapping code that only makes sense to be called once anyway.

**Reply by tcharding** [reply]
The problem the macros attempt to solve can be described by this series of events:
 
- In some `bitcoind` version an api call is introduced
- It remains for a couple of versions then the data it returns changes 
- At some stage it is removed

This then happens a whole bunch of different times for different api methods. The solution I came up with (eventually) was defining every API method (and test) in single macro so that

- Each client just defines the api methods they need using the appropriately named macro
- Each client uses the macro version that returns the correct data (or implements the test as needed).

Does that make sense?

**Reply by tcharding** [reply]
And the bit I'm really happy about is that because of the naming scheme using `__methodname` we should be able to write some simple shell scripts that parse out exactly what methods are supported for which versions - if that is useful.

One problem is that test fails don't print the line where the failure is and `track_caller` didn't help.

**Reply by tcharding** [reply]
Thanks for looking @apoelstra!

**Comment by tcharding** [comment]
I've got it to a state that `bitcoind-json-rpc` can use @RCasatta's `bitcoind` crate to run `bitcoind` instances in the integration tests, while the `bitcoind` crate depends on `bitcoin-json-rpc` for version specific RPC clients in its unit tests (currently only works with `v0.17.1` and `v22.1`).

- https://github.com/tcharding/rust-bitcoind-json-rpc
- https://github.com/rust-bitcoin/bitcoind/pull/164

**Comment by tcharding** [comment]
I'd like to proceed as follows:

1. Release `rust-bitcoind-json-rpc` from my own github account.
2. Attempt to patch `miniscript` to use it. 
 
That solves our release problem (`bitcoin`, `miniscript`, `bitcoind`, `bitcoincore-rpc-json`), I'll make sure the new crate is tested and ready to release as part of the `bitcoin` RC cycle.

3. Attempt to implement every Core JSON-RPC API method, that is a lot of grind work.
4. Attempt to move the crate either to `github.com/rust-bitcoin` or into the `rust-bitcoincore-rpc` repo, depending on @stevenroose's position.

At this stage, if anyone (`electrs`, `fedmint`) wants to use it, they can or not at their own discretion.

**Reply by apoelstra** [reply]
Sounds good, ACK steps 1 and 2.

**Reply by tcharding** [reply]
I just saw the bitcoin optech advice to be running Core 25, does that mean I'm totally wasting my time supporting all the old versions? After the vulnerability drops is no one ever going to run old Core versions any more?

**Reply by storopoli** [reply]
Whatever comes out of this "which version of Core should we officially support/test" (which is in my opinion a very important discussion) should be documented in `SECURITY.md`.

**Reply by apoelstra** [reply]
> I just saw the bitcoin optech advice to be running Core 25, does that mean I'm totally wasting my time supporting all the old versions? After the vulnerability drops is no one ever going to run old Core versions any more?

Lol, no. People will be running old Core versions forever. And if they're doing it only on their local network, using a modern version to bridge to the Internet, they'll be fine, security-wise.

For RPC purposes we ought to support as many versions as we can.

**Reply by tcharding** [reply]
Oh phew, I was worried I had wasted my time. Is there a reason to go back further that 0.17.1? The only reason behind that choice as earliest supported version is that its what was already done in `bitcoincore-rpc` and `bitcoind` (assumedly the second was just based on the first though).

**Reply by apoelstra** [reply]
>Is there a reason to go back further that 0.17.1?

Probably not. It's six years old now and [looks like it made a bunch of RPC changes that'd be a bunch of work for us to undo](https://bitcoincore.org/en/releases/0.17.0/).

Though we should make sure it's possible for people to hook up whatever version of Core they want and just lie to our library about what version it is. Most of the "basic" RPCs should technically work back to 0.3.0 or even earlier.

**Reply by tcharding** [reply]
Yep, at the moment I'm just focusing delivering a client that is for integration testing. If I've got motivation after that I'll see if we want to provide a client for production usage. I'm not sure at the moment how it would look and if I want to maintain it.

**Comment by cornbread78** [comment]
I'm Satoshi Nakamoto

**Reply by tcharding** [reply]
I blocked this user from the org, is that ok @apoelstra?

**Reply by apoelstra** [reply]
Yep, for sure.
